src/include/txn.h

```diff
@@ -177,13 +177,26 @@ struct __wt_txn_op {
 		WT_TXN_OP_NONE,
 		WT_TXN_OP_BASIC,
 		WT_TXN_OP_INMEM,
+		WT_TXN_OP_BASIC_ROW,
+		WT_TXN_OP_INMEM_ROW,
 		WT_TXN_OP_REF_DELETE,
 		WT_TXN_OP_TRUNCATE_COL,
 		WT_TXN_OP_TRUNCATE_ROW
 	} type;
 	union {
-		/* WT_TXN_OP_BASIC, WT_TXN_OP_INMEM */
-		WT_UPDATE *upd;
+		/*
+		 * WT_TXN_OP_BASIC, WT_TXN_OP_BASIC_ROW,
+		 * WT_TXN_OP_INMEM, WT_TXN_OP_INMEM_ROW
+		 */
+		struct {
+			WT_UPDATE *upd;
+			union {
+				/* WT_TXN_OP_BASIC_ROW, WT_TXN_OP_INMEM_ROW */
+				WT_ITEM row_key;
+				/* WT_TXN_OP_BASIC, WT_TXN_OP_INMEM */
+				uint64_t recno;
+			} key; /* Key to search prepared update. */
```

Comment:
Each time a new TXN_OP is created the key needs to be copied in, so the code will save both the key and update structure. That's done because there is no obvious way to get the from the TXN_OP or WT_UPDATE structure at prepare time.